File: mt5api/handlers/terminal.py
import MetaTrader5 as mt5
from flask import jsonify
from mt5api.config import MODE, UTC_OFFSET_HOURS, UTC_OFFSET_SECONDS
from mt5api.mt5client import (
    ensure_initialized,
    m,
    restart_terminal,
    to_dict,
    with_mt5,
)


def ping():
    return jsonify({"status": "ok", "mode": MODE})


@with_mt5
def last_error():
    err = m(mt5.last_error)
    return jsonify({"code": err[0], "message": err[1]})


@with_mt5
def get_terminal():
    if not ensure_initialized():
        return jsonify({"error": "MT5 not initialized"}), 503
    info = to_dict(m(mt5.terminal_info))
    if info is not None:
        info["broker_utc_offset_hours"] = UTC_OFFSET_HOURS
        info["broker_utc_offset_seconds"] = UTC_OFFSET_SECONDS
    return jsonify(info)


@with_mt5
def init():
    if not ensure_initialized():
        err = m(mt5.last_error)
        return jsonify({"success": False, "error": str(err)}), 500
    return jsonify({"success": True})


@with_mt5
def shutdown():
    m(mt5.shutdown)
    return jsonify({"success": True})


@with_mt5
def restart():
    if not restart_terminal():
        return jsonify({"success": False, "error": "Terminal restart failed"}), 500
    return jsonify({"success": True})

File: clients/go/terminal.go
package mt5httpapi

import (
	"context"
	"net/http"

	"github.com/psyb0t/ctxerrors"
)

func (c *Client) Ping(ctx context.Context) (*PingResponse, error) {
	out := &PingResponse{}
	if err := c.do(ctx, http.MethodGet, "/ping", nil, nil, out); err != nil {
		return nil, ctxerrors.Wrap(err, "ping")
	}

	return out, nil
}

func (c *Client) LastError(ctx context.Context) (*LastError, error) {
	out := &LastError{}
	if err := c.do(ctx, http.MethodGet, "/error", nil, nil, out); err != nil {
		return nil, ctxerrors.Wrap(err, "get last error")
	}

	return out, nil
}

func (c *Client) GetTerminal(ctx context.Context) (*Terminal, error) {
	out := &Terminal{}
	if err := c.do(ctx, http.MethodGet, "/terminal", nil, nil, out); err != nil {
		return nil, ctxerrors.Wrap(err, "get terminal")
	}

	return out, nil
}

func (c *Client) InitTerminal(ctx context.Context) error {
	if err := c.do(ctx, http.MethodPost, "/terminal/init", nil, nil, nil); err != nil {
		return ctxerrors.Wrap(err, "init terminal")
	}

	return nil
}

func (c *Client) ShutdownTerminal(ctx context.Context) error {
	if err := c.do(ctx, http.MethodPost, "/terminal/shutdown", nil, nil, nil); err != nil {
		return ctxerrors.Wrap(err, "shutdown terminal")
	}

	return nil
}

func (c *Client) RestartTerminal(ctx context.Context) error {
	if err := c.do(ctx, http.MethodPost, "/terminal/restart", nil, nil, nil); err != nil {
		return ctxerrors.Wrap(err, "restart terminal")
	}

	return nil
}
